### PR TITLE
LibJS: Implement normative Temporal spec changes

### PR DESCRIPTION
**Pull request by linusg**

### CODE REVIEW

File: Userland/Libraries/LibJS/Runtime/Temporal/CalendarPrototype.cpp [hunk 1]
```diff
@@ -66,10 +66,11 @@ void CalendarPrototype::initialize(GlobalObject& global_object)
 JS_DEFINE_NATIVE_FUNCTION(CalendarPrototype::id_getter)
 {
     // 1. Let calendar be the this value.
-    auto calendar = vm.this_value(global_object);
+    // 2. Perform ? RequireInternalSlot(calendar, [[InitializedTemporalCalendar]]).
+    auto* calendar = TRY(typed_this_object(global_object));
 
     // 2. Return ? ToString(calendar).
```
Comment (IdanHo) [inline Userland/Libraries/LibJS/Runtime/Temporal/CalendarPrototype.cpp:72]: 3.

File: Userland/Libraries/LibJS/Runtime/Temporal/PlainDatePrototype.cpp [hunk 2]
```diff
@@ -461,22 +461,25 @@ JS_DEFINE_NATIVE_FUNCTION(PlainDatePrototype::until)
...
 
-    // 12. Let untilOptions be ? MergeLargestUnitOption(options, largestUnit).
+    // 13. Let untilOptions be ? MergeLargestUnitOption(options, largestUnit).
     auto* until_options = TRY(merge_largest_unit_option(global_object, *options, move(*largest_unit)));
 
-    // 13. Let result be ? CalendarDateUntil(temporalDate.[[Calendar]], temporalDate, other, untilOptions).
+    // 14. Let result be ? CalendarDateUntil(temporalDate.[[Calendar]], temporalDate, other, untilOptions).
     auto* result = TRY(calendar_date_until(global_object, temporal_date->calendar(), temporal_date, other, *until_options));
 
     // NOTE: Result can be reassigned by 14.b, `result` above has the type `Duration*` from calendar_date_until while 14.b has the type `RoundedDuration` from round_duration.
```
Comment (IdanHo) [inline Userland/Libraries/LibJS/Runtime/Temporal/PlainDatePrototype.cpp:485]: 14.b is no longer a thing (14 is now 15 and b doesn't exist)

File: Userland/Libraries/LibJS/Runtime/Temporal/AbstractOperations.cpp [hunk 3]
```diff
@@ -1147,13 +1147,13 @@ ThrowCompletionOr<ISODateTime> parse_iso_date_time(GlobalObject& global_object,
         auto fraction = String::formatted("{}000000000", *fraction_part);
         // b. Let millisecond be the String value equal to the substring of fraction from 0 to 3.
```
Comment (IdanHo) [inline Userland/Libraries/LibJS/Runtime/Temporal/AbstractOperations.cpp:1148]: These comments should be updated as well

File: Userland/Libraries/LibJS/Runtime/Temporal/ISO8601.cpp [hunk 4]
```diff
@@ -819,6 +808,17 @@ bool ISO8601Parser::parse_date_time()
...
 
+// https://tc39.es/proposal-temporal/#prod-CalendarTime
+bool ISO8601Parser::parse_calendar_time()
+{
+    // CalendarTime :
+    //     TimeSpec TimeZone[opt] Calendar[opt]
+    if (!parse_time_spec())
+        return false;
+    (void)parse_time_zone();
+    return true;
```
Comment (IdanHo) [inline Userland/Libraries/LibJS/Runtime/Temporal/ISO8601.cpp:822]: Missing calendar parse?

File: Userland/Libraries/LibJS/Runtime/Temporal/ISO8601.cpp [hunk 5]
```diff
@@ -1173,23 +1173,23 @@ bool ISO8601Parser::parse_temporal_month_day_string()
...
 
 // https://tc39.es/proposal-temporal/#prod-TemporalTimeString
 bool ISO8601Parser::parse_temporal_time_string()
 {
     // TemporalTimeString :
-    //     Time
-    //     DateTime
+    //     CalendarTime
+    //     CalendarDateTime
     // NOTE: Reverse order here because `Time` can be a subset of `DateTime`,
```
Comment (IdanHo) [inline Userland/Libraries/LibJS/Runtime/Temporal/ISO8601.cpp:1208]: Comment should be updated to CalendarDateTime

File: Userland/Libraries/LibJS/Runtime/Temporal/ISO8601.cpp [hunk 6]
```diff
@@ -1226,10 +1226,10 @@ bool ISO8601Parser::parse_temporal_year_month_string()
 {
     // TemporalYearMonthString :
     //     DateSpecYearMonth
-    //     DateTime
+    //     CalendarDateTime
     // NOTE: Reverse order here because `DateSpecYearMonth` can be a subset of `DateTime`,
```
Comment (IdanHo) [inline Userland/Libraries/LibJS/Runtime/Temporal/ISO8601.cpp:1249]: Same here

File: Userland/Libraries/LibJS/Runtime/Temporal/ISO8601.cpp [hunk 7]
```diff
@@ -833,6 +833,22 @@ bool ISO8601Parser::parse_calendar_date_time()
     return true;
 }
 
+// https://tc39.es/proposal-temporal/#prod-CalendarDateTimeTimeRequired
+bool ISO8601Parser::parse_calendar_date_time_time_required()
```
Comment (IdanHo) [inline Userland/Libraries/LibJS/Runtime/Temporal/ISO8601.cpp:837]: Do we need to actually use this somewhere?

**Review by IdanHo — COMMENTED** [review]
Nice, just a couple of (mostly comment-related) comments

**Comment by linusg** [comment]
Thanks, all fixed!

**Comment by IdanHo** [comment]
Commit `24c92bfa3c6e426713930a47c17254d2c8a1dee6` is still missing the calendar parse :sweat_smile:

**Comment by linusg** [comment]
um, no? Line 821:

image

**Comment by linusg** [comment]
(sorry, that commit is gone now as I force pushed some more comment tweaks)

**Comment by IdanHo** [comment]
image
[`a45f5c6` (#11298)](https://github.com/SerenityOS/serenity/pull/11298/commits/a45f5c67163261e0efaf1642d55fd2423d3b430a)

**Comment by linusg** [comment]
:facepalm:

sorry about that, rebasing is hard :D

**Review by IdanHo — APPROVED** [review]

**Pull request merged**